File: Sources/Ably/ARTJsonEncoder.m
//
//  ARTJsonEncoder.m
//
//  Created by Jason Choy on 09/12/2014.
//  Copyright (c) 2014 Ably. All rights reserved.
//

#import "ARTJsonEncoder.h"

@implementation ARTJsonEncoder

- (NSString *)mimeType {
    return @"application/json";
}

- (ARTEncoderFormat)format {
    return ARTEncoderFormatJson;
}

- (NSString *)formatAsString {
    return @"json";
}

- (id)decode:(NSData *)data error:(NSError **)error {
    return [NSJSONSerialization JSONObjectWithData:data options:0 error:error];
}

- (NSData *)encode:(id)obj error:(NSError **)error {
    @try {
        NSJSONWritingOptions options;
        if (@available(macOS 10.13, iOS 11.0, tvOS 11.0, *)) {
            options = NSJSONWritingSortedKeys;
        }
        else {
            options = 0;
        }
        return [NSJSONSerialization dataWithJSONObject:obj options:options error:error];
    }
    @catch (NSException *exception) {
        if (error) {
            *error = [[NSError alloc] initWithDomain:ARTAblyErrorDomain code:ARTClientCodeErrorInvalidType userInfo:@{NSLocalizedDescriptionKey: exception.reason}];
        }
        return nil;
    }
}

@end

File: Sources/AblySwift/Core/ARTJsonEncoder.swift
import Foundation

public class ARTJsonEncoder: ARTJsonLikeEncoderDelegate {
    
    public func mimeType() -> String {
        return "application/json"
    }
    
    public func format() -> ARTEncoderFormat {
        return .json
    }
    
    public func formatAsString() -> String {
        return "json"
    }
    
    public func decode(_ data: Data) throws -> Any? {
        return try JSONSerialization.jsonObject(with: data, options: [])
    }
    
    public func encode(_ obj: Any) throws -> Data? {
        do {
            var options: JSONSerialization.WritingOptions = []
            if #available(macOS 10.13, iOS 11.0, tvOS 11.0, *) {
                options = .sortedKeys
            }
            return try JSONSerialization.data(withJSONObject: obj, options: options)
        } catch {
            // Convert to Ably error format
            let ablyError = NSError(domain: ARTAblyErrorDomain, 
                                   code: ARTClientCodeErrorInvalidType, 
                                   userInfo: [NSLocalizedDescriptionKey: error.localizedDescription])
            throw ablyError
        }
    }
}

// Error codes
public let ARTClientCodeErrorInvalidType = 40000